tests/fuzz/pki_fuzzer.go

```diff
@@ -0,0 +1,222 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// nolint: golint
+package fuzz
+
+import (
+	"crypto/x509/pkix"
+	"os"
+
+	fuzz "github.com/AdaLogics/go-fuzz-headers"
+
+	"istio.io/istio/security/pkg/pki/util"
+)
+
+// FuzzVerifyCertificate implements a fuzzer
+// that tests util.VerifyCertificate
+func FuzzVerifyCertificate(data []byte) int {
+	f := fuzz.NewConsumer(data)
+	privPem, err := f.GetBytes()
+	if err != nil {
+		return 0
+	}
+	certChainPem, err := f.GetBytes()
+	if err != nil {
+		return 0
+	}
+	rootCertPem, err := f.GetBytes()
+	if err != nil {
+		return 0
+	}
+	expectedFields := &util.VerifyFields{}
+	err = f.GenerateStruct(expectedFields)
+	if err != nil {
+		return 0
+	}
+	util.VerifyCertificate(privPem, certChainPem, rootCertPem, expectedFields)
+	return 1
+}
+
+// FindRootCertFromCertificateChainBytesFuzz implements a fuzzer
+// that tests util.FindRootCertFromCertificateChainBytes
+func FindRootCertFromCertificateChainBytes(data []byte) int {
+	_, _ = util.FindRootCertFromCertificateChainBytes(data)
+	return 1
+}
+
+func FuzzExtractIDs(data []byte) int {
+	f := fuzz.NewConsumer(data)
+	noOfExts, err := f.GetInt()
+	if err != nil {
+		return 0
+	}
+	noOfExtensions := noOfExts % 30
+	if noOfExtensions == 0 {
+		return 0
+	}
+	extensions := make([]pkix.Extension, 0)
+	for i := 0; i < noOfExtensions; i++ {
+		newExtension := pkix.Extension{}
+		err = f.GenerateStruct(&newExtension)
+		if err != nil {
+			return 0
+		}
+		extensions = append(extensions, newExtension)
+	}
+	_, _ = util.ExtractIDs(extensions)
+	return 1
+}
+
+// FuzzPemCertBytestoString implements a fuzzer
+// that tests PemCertBytestoString
+func FuzzPemCertBytestoString(data []byte) int {
+	_ = util.PemCertBytestoString(data)
+	return 1
+}
+
+// FuzzParsePemEncodedCertificateChain implements
+// a fuzzer that tests ParsePemEncodedCertificateChain
+func FuzzParsePemEncodedCertificateChain(data []byte) int {
+	_, _ = util.ParsePemEncodedCertificateChain(data)
+	return 1
+}
+
+// FuzzUpdateVerifiedKeyCertBundleFromFile implements a
+// fuzzer that tests UpdateVerifiedKeyCertBundleFromFile
+func FuzzUpdateVerifiedKeyCertBundleFromFile(data []byte) int {
+	f := fuzz.NewConsumer(data)
+	certFile, err := os.Create("certfile")
```

Comment:
can we delete it on cleanup?

Also is using files for fuzzing really appropriate? Seems like we should just fuzz NewVerifiedKeyCertBundleFromPem instead?

Reply:
> can we delete it on cleanup?

Sure, added.

> Also is using files for fuzzing really appropriate?

Yes, it is fine as long as it is fast enough. I suggest we try `UpdateVerifiedKeyCertBundleFromFile` and observe if the speed is poor after a few weeks. Otherwise we stick with `NewVerifiedKeyCertBundleFromPem`.